proposals/20230316-oci-content-source.md

```diff
@@ -0,0 +1,284 @@
+The following is an example image manifest for a component containing a configuration structure, a layer containing the `my-component.wasm` binary.
+```json
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "artifactType": "application/vnd.wasm.component.v1",
+  "config": {
+    "mediaType": "application/vnd.wasm.component.config.v1+json",
+    "digest": "sha256:5587da2246a78f08c447bff2ac91ee5c2b57be2f2a15244b5e618ac0be626885",
+    "size": 331
+  },
+  "layers": [
+    {
+      "mediaType": "application/vnd.wasm.content.layer.v1+wasm",
+      "digest": "sha256:b36aa5d0111a488937361fdb35432510d50675a11d566c5d9e82a147fb9ff552",
+      "size": 2087464,
+      "annotations": {
+        "org.opencontainers.image.title": "my-component"
+      }
+    }
+  ]
+}
+```
+
+The following is an example of the configuration structure referenced in the preceding image manifest.
+```json
+{
+  "mediaType": "application/vnd.wasm.component.config.v1+json",
+  "architecture": "wasm32",
+  "os": "wasi"
+}
+```
+
+#### Image Manifest for Interface Components
+The following is an example image manifest for a component containing a configuration structure and a layer containing the `my-component-interface.wasm` binary.
+```json
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "artifactType": "application/vnd.wasm.component.interface.v1",
+  "config": {
+    "mediaType": "application/vnd.wasm.component.config.v1+json",
+    "digest": "sha256:c71d239df91726fc519c6eb72d318ec65820627232b2f796219e87dcf35d0ab4",
+    "size": 331
+  },
+  "layers": [
+    {
+      "mediaType": "application/vnd.wasm.content.layer.v1+wasm",
+      "digest": "sha256:dcf07c6bb395d6e1d40505b77e70af04f5fae0d54c9573fd379c1e7355a18cf3",
+      "size": 2087464,
+      "annotations": {
+        "org.opencontainers.image.title": "my-component-interface"
+      }
+    }
+  ]
+}
+```
+
+The following is an example of the configuration structure referenced in the preceding image manifest.
+```json
+{
+  "mediaType": "application/vnd.wasm.component.config.v1+json",
+  "architecture": "wasm32",
+  "os": "wasi"
+}
+```
+
+#### Image Manifest for Bundled Components
+The following is an example image manifest for a component containing a configuration structure, a layer containing the `my-bundled-component.wasm` binary, and a data layer containing some a static asset.
+```json
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "artifactType": "application/vnd.wasm.component.bundled.v1",
+  "config": {
+    "mediaType": "application/vnd.wasm.component.config.v1+json",
+    "digest": "sha256:105ab3237b4f0d885700892a0f4b3482d1146dff27c88d46f02b8bd7ef67c3de",
+    "size": 331
+  },
+  "layers": [
+    {
+      "mediaType": "application/vnd.wasm.content.layer.v1+wasm",
+      "digest": "sha256:2e94e0582fb925e89515435513496819dc8f364f2da400059a64d6d1412ca2ad",
+      "size": 2087464,
```

Comment:
Is the assumption here that the bundled component is a statically linked component that includes all composing components and core modules?

This means there is no way to do layer de-duplication when pushing, and the registry would have to store the bytes for the same components for every component that takes a certain component as a dependency.

I'd like to propose a transparent way for clients to split bundled component at distribution time, and reassemble it when pulling the artifact.
This is described in this proposal — https://hackmd.io/50rfwV6BTJWN8VZBhdAN_g

And a prototype for such a tool can be found here — https://github.com/fermyon/wasm-splice


We could take this approach for nested components, core modules, and data sections.